cpp/src/arrow/ipc/ipc-adapter-test.cc

```diff
@@ -0,0 +1,109 @@
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <gtest/gtest.h>
+#include <cstdint>
+#include <cstdio>
+#include <cstring>
+#include <memory>
+#include <random>
+#include <string>
+#include <vector>
+
+#include "arrow/ipc/adapter.h"
+#include "arrow/ipc/memory.h"
+#include "arrow/ipc/test-common.h"
+
+#include "arrow/test-util.h"
+#include "arrow/types/primitive.h"
+#include "arrow/util/bit-util.h"
+#include "arrow/util/buffer.h"
+#include "arrow/util/memory-pool.h"
+#include "arrow/util/status.h"
+
+namespace arrow {
+namespace ipc {
+
+class TestWriteRowBatch : public ::testing::Test, public MemoryMapFixture {
+ public:
+  void SetUp() {
+    pool_ = default_memory_pool();
+  }
+  void TearDown() {
+    MemoryMapFixture::TearDown();
+  }
+
+  void InitMemoryMap(int64_t size) {
+    std::string path = "test-write-row-batch";
+    MemoryMapFixture::CreateFile(path, size);
+    ASSERT_OK(MemoryMappedSource::Open(path, MemorySource::READ_WRITE, &mmap_));
+  }
+
+ protected:
+  MemoryPool* pool_;
+  std::shared_ptr<MemoryMappedSource> mmap_;
+};
+
+const auto INT32 = std::make_shared<Int32Type>();
+
+TEST_F(TestWriteRowBatch, IntegerRoundTrip) {
+  int length = 1000;
+
+  // Make the schema
+  auto f0 = std::make_shared<Field>("f0", INT32);
+  auto f1 = std::make_shared<Field>("f1", INT32);
+  std::shared_ptr<Schema> schema(new Schema({f0, f1}));
+
+  // Example data
+
+  auto data = std::make_shared<PoolBuffer>(pool_);
+  ASSERT_OK(data->Resize(length * sizeof(int32_t)));
+  test::rand_uniform_int(length, 0, 0, 1000,
+      reinterpret_cast<int32_t*>(data->mutable_data()));
+
+  auto nulls = std::make_shared<PoolBuffer>(pool_);
+  int null_bytes = util::bytes_for_bits(length);
+  ASSERT_OK(nulls->Resize(null_bytes));
+  test::random_bytes(null_bytes, 0, nulls->mutable_data());
+
+  auto a0 = std::make_shared<Int32Array>(length, data);
+  auto a1 = std::make_shared<Int32Array>(length, data,
+      test::bitmap_popcount(nulls->data(), length), nulls);
+
+  RowBatch batch(schema, length, {a0, a1});
+
+  // TODO(wesm): computing memory requirements for a row batch
+  // 64k is plenty of space
+  InitMemoryMap(1 << 16);
+
+  int64_t header_location;
+  ASSERT_OK(WriteRowBatch(mmap_.get(), &batch, 0, &header_location));
+
+  std::shared_ptr<RowBatchReader> result;
+  ASSERT_OK(RowBatchReader::Open(mmap_, header_location, &result));
+
+  std::shared_ptr<RowBatch> batch_result;
+  ASSERT_OK(result->GetRowBatch(schema, &batch_result));
+  ASSERT_EQ(batch.num_rows(), batch_result->num_rows());
+
+  for (int i = 0; i < batch.num_columns(); ++i) {
+    ASSERT_TRUE(batch.column(i)->Equals(batch_result->column(i))) << i;
```

Comment:
EXPECT_TRUE might be more appropriate here, it would allow the user to get a full report if something fails...